qiskit/dagcircuit/dagcanonical.py

```diff
@@ -0,0 +1,400 @@
+        return self.to_networkx().nodes[node_id]['successors']
+
+    def predecessors(self, node_id):
+        """Returns set of the descendants of a node as DAGNodes."""
+        return self.to_networkx().nodes[node_id]['predecessors']
+
+    def draw(self, scale=0.7, filename=None, style='color', category='canonical'):
+        """
+        Draws the DAG canonical circuit.
+
+        This function needs `pydot <https://github.com/erocarrera/pydot>`, which in turn needs
+        Graphviz <https://www.graphviz.org/>` to be installed.
+
+        Args:
+            scale (float): scaling factor
+            filename (str): file path to save image to (format inferred from name)
+            style (str): 'plain': B&W graph
+                         'color' (default): color input/output/op nodes
+            category(str): 'canonical' Other type of DAG
+
+        Returns:
+            Ipython.display.Image: if in Jupyter notebook and not saving to file,
+                otherwise None.
+        """
+        from qiskit.visualization.dag_visualization import dag_drawer
+        return dag_drawer(dag=self, scale=scale, filename=filename, style=style, category=category)
+
+
+def merge_no_duplicates(*iterables):
+    """Merge K list without duplicate using python heapq ordered merging
+
+    Args:
+        *iterables: A list of k sorted lists
+
+    Yields:
+        Iteraor: List from the merging of the k ones (without duplicates
+    """
+    last = object()
+    for val in heapq.merge(*iterables):
+        if val != last:
+            last = val
+            yield val
+
+
+def _commute(node1, node2):
+    """Function to verify commutation relation between two nodes in the DAG
+
+    Args:
+        node1 (DAGnode): first node operation (attribute ['operation'] in the DAG)
+        node2 (DAGnode): second node operation
+
+    Return:
+        bool: True if the gates commute and false if it is not the case.
+    """
+
+    # Create set of qubits on which the operation acts
+    qarg1 = [node1.qargs[i].index for i in range(0, len(node1.qargs))]
+    qarg2 = [node2.qargs[i].index for i in range(0, len(node2.qargs))]
+
+    # Create set of cbits on which the operation acts
+    carg1 = [node1.qargs[i].index for i in range(0, len(node1.cargs))]
+    carg2 = [node2.qargs[i].index for i in range(0, len(node2.cargs))]
+
+    # Commutation for classical conditional gates
+    if node1.condition or node2.condition:
+        intersection = set(qarg1).intersection(set(qarg2))
+        if intersection or carg1 or carg2:
+            commute_condition = False
+        else:
+            commute_condition = True
+        return commute_condition
+
+    # Commutation for measurement
+    if node1.name == 'measure' or node2.name == 'measure':
+        intersection = set(qarg1).intersection(set(qarg2))
+        if intersection:
+            commute_measurement = False
+        else:
+            commute_measurement = True
```

Comment:
if two measurements are on different qubits, but into the same classical bits, this says they commute. But I think they don't.